File: src/web/banks/registry.js
/** Registry of bank statement extractors. */

import { tdCreditExtractor } from "./td_credit.js";

const EXTRACTORS = {
  [tdCreditExtractor.id]: tdCreditExtractor,
};

export function listExtractors() {
  return Object.values(EXTRACTORS);
}

export function getExtractor(bankId) {
  const extractor = EXTRACTORS[bankId];
  if (!extractor) {
    const known = Object.keys(EXTRACTORS).sort().join(", ");
    throw new Error(`Unknown bank: ${bankId}. Known banks: ${known}`);
  }
  return extractor;
}

export function detectExtractor(pages) {
  for (const extractor of Object.values(EXTRACTORS)) {
    if (extractor.detect(pages)) {
      return extractor;
    }
  }
  return null;
}

File: src/banks/registry.py
"""Registry of bank statement extractors."""

from __future__ import annotations

from typing import Any

from banks.base import BankExtractor
from banks.td_credit import TdCreditExtractor

_EXTRACTORS: dict[str, BankExtractor] = {
    TdCreditExtractor.id: TdCreditExtractor(),
}


def list_extractors() -> list[BankExtractor]:
    return list(_EXTRACTORS.values())


def get_extractor(bank_id: str) -> BankExtractor:
    extractor = _EXTRACTORS.get(bank_id)
    if extractor is None:
        known = ", ".join(sorted(_EXTRACTORS))
        raise ValueError(f"Unknown bank: {bank_id!r}. Known banks: {known}")
    return extractor


def detect_extractor(pages: list[dict[str, Any]]) -> BankExtractor | None:
    for extractor in _EXTRACTORS.values():
        if extractor.detect(pages):
            return extractor
    return None
